src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -257,91 +317,103 @@ public boolean isFingerprintArtifacts() {
+        }
+
+        EnvVars env;
+        if (build instanceof AbstractBuild) {
+            env = build.getEnvironment(listener);
+            env.overrideAll(((AbstractBuild)build).getBuildVariables()); // Add in matrix axes..
+        } else {
+            env = new EnvVars();
+        }
+
         PrintStream console = listener.getLogger();
         String expandedProject = project, expandedFilter = filter;
         String expandedExcludes = getExcludes();
-        try {
-            EnvVars env = build.getEnvironment(listener);
-            env.overrideAll(build.getBuildVariables()); // Add in matrix axes..
-            expandedProject = env.expand(project);
-            Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, build.getProject().getRootProject().getParent(), Job.class);
-            if (job != null && !expandedProject.equals(project)
-                // If projectName is parameterized, need to do permission check on source project.
-                && !canReadFrom(job, build)) {
-                job = null; // Disallow access
-            }
-            if (job == null) {
-                console.println(Messages.CopyArtifact_MissingProject(expandedProject));
-                return false;
-            }
-            Run src = selector.getBuild(job, env, parameters != null ? new ParametersBuildFilter(env.expand(parameters)) : new BuildFilter(), build);
-            if (src == null) {
-                console.println(Messages.CopyArtifact_MissingBuild(expandedProject));
-                return isOptional();  // Fail build unless copy is optional
+
+        expandedProject = env.expand(project);
+        Job<?, ?> job = Jenkins.getInstance().getItem(expandedProject, getItemGroup(build), Job.class);
+        if (job != null && !expandedProject.equals(project)
+            // If projectName is parameterized, need to do permission check on source project.
+            && !canReadFrom(job, build)) {
+            job = null; // Disallow access
+        }
+        if (job == null) {
+            throw new AbortException(Messages.CopyArtifact_MissingProject(expandedProject));
+        }
+        Run src = selector.getBuild(job, env, parameters != null ? new ParametersBuildFilter(env.expand(parameters)) : new BuildFilter(), build);
+        if (src == null) {
+            String message = Messages.CopyArtifact_MissingBuild(expandedProject);
+            if (isOptional()) {
+                // just return without an error
+                console.println(message);
+                return;
+            } else {
+                // Fail build if copy is not optional
+                throw new AbortException(message);
             }
-            FilePath targetDir = build.getWorkspace(), baseTargetDir = targetDir;
-            if (targetDir == null || !targetDir.exists()) {
-                console.println(Messages.CopyArtifact_MissingWorkspace()); // (see JENKINS-3330)
```

Comment:
Assuming this was the only usage, you can now delete the key from `Messages.properties`.